extensions/markdown/src/extension.ts

```diff
@@ -0,0 +1,170 @@
+}
+
+function getMarkdownUri(document: vscode.TextDocument) {
+	return document.uri.with({ scheme: 'markdown', query: document.uri.toString() });
+}
+
+function openPreview(sideBySide?: boolean): void {
+	const activeEditor = vscode.window.activeTextEditor;
+	if (!activeEditor) {
+		vscode.commands.executeCommand('workbench.action.navigateBack');
+		return;
+	}
+
+	let markdownPreviewUri = getMarkdownUri(activeEditor.document);
+	vscode.commands.executeCommand('vscode.previewHtml',
+		markdownPreviewUri,
+		getViewColumn(sideBySide),
+		`Preview '${path.basename(activeEditor.document.fileName)}'`);
+}
+
+function getViewColumn(sideBySide): ViewColumn {
+	const active = vscode.window.activeTextEditor;
+	if (!active) {
+		return ViewColumn.One;
+	}
+
+	if (!sideBySide) {
+		return active.viewColumn;
+	}
+
+	switch (active.viewColumn) {
+		case ViewColumn.One:
+			return ViewColumn.Two;
+		case ViewColumn.Two:
+			return ViewColumn.Three;
+	}
+
+	return active.viewColumn;
+}
+
+
+class MDDocumentContentProvider implements TextDocumentContentProvider {
+	private _context: ExtensionContext;
+	private _onDidChange = new EventEmitter<Uri>();
+
+	constructor(context: ExtensionContext) {
+		this._context = context;
+	}
+
+	private getMediaPath(mediaFile) {
+		return this._context.asAbsolutePath(path.join('media', mediaFile));
+	}
+
+	private fixHref(resource: Uri, href: string) {
+		if (href) {
+			// Return early if href is already a URL
+			if (Uri.parse(href).scheme) {
+				return href;
+			}
+			// Otherwise convert to a file URI by joining the href with the resource location
+			return Uri.file(path.join(path.dirname(resource.fsPath), href)).toString();
+		}
+		return href;
+	}
+
+	private computeCustomStyleSheetIncludes(uri: Uri): string[] {
+		const styles = vscode.workspace.getConfiguration('markdown')['styles'];
+		if (styles && Array.isArray(styles)) {
+			return styles.map((style) => {
+				return `<link rel="stylesheet" href="${this.fixHref(uri, style)}" type="text/css" media="screen">`;
+			});
+		}
+		return [];
+	}
+
+	public provideTextDocumentContent(uri: Uri): Thenable<string> {
+
+		return vscode.workspace.openTextDocument(Uri.parse(uri.query)).then(document => {
+			const head = [].concat(
+				'<!DOCTYPE html>',
+				'<html>',
+				'<head>',
+				'<meta http-equiv="Content-type" content="text/html;charset=UTF-8">',
+				`<link rel="stylesheet" type="text/css" href="${this.getMediaPath('markdown.css')}" >`,
+				`<link rel="stylesheet" type="text/css" href="${this.getMediaPath('tomorrow.css')}" >`,
+				this.computeCustomStyleSheetIncludes(uri),
+				'</head>',
+				'<body>'
+			).join('\n');
+
+			const body = md.render(document.getText());
+
+			const tail = [
+				'</body>',
+				'</html>'
+			].join('\n');
+
+			return head + body + tail;
+		});
+	}
+
+	get onDidChange(): Event<Uri> {
+		return this._onDidChange.event;
+	}
+
+	public update(uri: Uri) {
+		this._onDidChange.fire(uri);
```

Comment:
I only really noticed a hit when I was trying out some heavier plugins like katex.  I was working on a similar plugin [here](https://github.com/cbreeden/vscode-markdownit/blob/master/src/extension.ts) yesterday before you brought this branch to my attention.  I noticed things like pasting multiple equations very quickly (something that you probably shouldn't do but I was curious how it would handle the stress) could cause the program to hang.  The issue really isn't saving too frequently, but typing quickly as markdown rendering is trigger on every document change here.

I'm going to go ahead and submit a pull request to [markdown-it-katex](https://github.com/search?utf8=%E2%9C%93&q=markdown-it-katex) to have an option to cache rendered formulas perhaps to see if that helps.  But ya, I'll go ahead an submit a PR if I notice a performance hit once the July release comes around.

Do note that the current markdown render [does have a timeout](https://github.com/Microsoft/vscode/blob/844889948d9ca957835452c8d6786c32dc79d991/src/vs/workbench/parts/markdown/browser/markdownExtension.ts#L36).  The difference here is that this extension is using TextDocumentContentProviders to handle events while the current one is doing it's own thing.